go/vt/wrangler/vexec.go

```diff
@@ -320,32 +356,57 @@ func (wr *Wrangler) getReplicationStatusFromRow(ctx context.Context, row []sqlty
 	return status, bls.Keyspace, nil
 }
 
-func (wr *Wrangler) getStreams(ctx context.Context, workflow, keyspace string) (*replicationStatusResult, error) {
-	var rsr replicationStatusResult
-	rsr.Statuses = make(map[string][]*replicationStatus)
+func (wr *Wrangler) getStreams(ctx context.Context, workflow, keyspace string) (*ReplicationStatusResult, error) {
+	var rsr ReplicationStatusResult
+	rsr.ShardStatuses = make(map[string]*ShardReplicationStatus)
 	rsr.Workflow = workflow
-	rsr.TargetKeyspace = keyspace
 	var results map[*topo.TabletInfo]*querypb.QueryResult
 	query := "select id, source, pos, stop_pos, max_replication_lag, state, db_name, time_updated, message from _vt.vreplication"
 	results, err := wr.runVexec(ctx, workflow, keyspace, query, false)
 	if err != nil {
 		return nil, err
 	}
+
+	// We set a topo timeout since we contact topo for the shard record.
+	ctx, cancel := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
+	defer cancel()
+	var sourceKeyspace string
+	sourceShards := sets.NewString()
+	targetShards := sets.NewString()
 	for master, result := range results {
-		var rsrStatus []*replicationStatus
+		var rsrStatus []*ReplicationStatus
 		qr := sqltypes.Proto3ToResult(result)
 		for _, row := range qr.Rows {
-			status, sourceKeyspace, err := wr.getReplicationStatusFromRow(ctx, row, master)
+			status, sk, err := wr.getReplicationStatusFromRow(ctx, row, master)
 			fmt.Printf("getReplicationStatusFromRow status for master %s is %v\n", master.AliasString(), status)
 			if err != nil {
 				return nil, err
 			}
-			rsr.SourceKeyspace = sourceKeyspace
+			sourceKeyspace = sk
+			sourceShards.Insert(status.Bls.Shard)
 
 			rsrStatus = append(rsrStatus, status)
 		}
-		rsr.Statuses[fmt.Sprintf("%s/%s", master.Shard, master.AliasString())] = rsrStatus
+		si, err := wr.ts.GetShard(ctx, keyspace, master.Shard)
```

Comment:
Adding a global topo call per shard could significantly change the latency and reliability of call sites. Each global topo call can take hundreds of ms to go around the world and back, and we have users with hundreds of shards. In addition, this will now fail when global topo is unavailable, whereas previously it may have still worked as long as cell-local topo is available, for example.

Is this new info always necessary for callers? If not, we should require them to fetch this information separately.

Reply:
This call was made for `ShowResharding` which this command intends to replace. I agree with you from a performance perspective, but on the flip side, this information is necessary for understanding the entire state of an ongoing resharding operation. We could force users to make this call separately, but that seems far from ideal from a usability perspective.

Reply:
The way I would answer this is.. can you successfully reshard if you can't reach global topo? If the answer is yes, then `Show` should also work. Otherwise, otherwise.

Reply:
I'm not sure what the answer is to that. @sougou?

If we leave out talking to global topo then there's a significant amount of information we won't be exposing about ongoing resharding, unless there's some other way to get the information?

Reply:
I just drilled down into vx.getMasters which gets called by every `VExec` and it's already doing a global topo call per shard @enisoc

Reply:
two global topo calls per shard in fact

Reply:
It gets all the info we need actually but throws most of it away. Unfortunately, everything has to happen through the barrier of `VExec`. I'm not sure how to have `VExec` return that info in a clean way so we don't have to make those calls again.

Reply:
I feel like some cleanup is in order, but can be a separate PR.

Reply:
Sorry, I didn't know `getStreams` is only used for user-facing information display, and not for the actual resharding data path. I take back this comment.